### DISCUSSION TITLE
Import a module with a helper function that returns it

### DISCUSSION THREAD

**Discussion by ClementPinard** [microsoft/pyright#5969] [Q&A] [answered]
Hi,

I have a library that uses [fiftyone](https://pypi.org/project/fiftyone/) as extra dependency. A known problem with fiftyone is the database package that depends on the linux distribution. See https://docs.voxel51.com/getting_started/troubleshooting.html#alternative-linux-builds

As such, I would like to import the fiftyone package in my library with a helper function that would raise an informative message on how to solve it.

Here is the type system that I use.

in `util.py`

```python
def try_import_fiftyone() -> types.ModuleType:
    """Try to import fiftyone and display an informative error

    Raises:
        ImportError: raise an error if the module could not be imported. It can be
            either because it's not installed, or because the database package
            (fyftone-db) is not compatible with the current distribution.

    Returns:
        types.ModuleType: the imported fiftyone module
    """
    try:
        import fiftyone as fo
    except ImportError as e:
        raise ImportError(
            "Fiftyone could not be loaded, make sure you have installed the library with the"
            " 'fiftyone' extra."
        ) from e
    except Exception as e:
        raise ImportError(
            "Fiftyone is installed but cannot be loaded, you might have to install a"
            " custom 'fyftone-db' for your OS. See"
            " https://docs.voxel51.com/getting_started/troubleshooting.html#alternative-linux-builds"
        ) from e
    return fo
```

in a submodule that uses fiftyone :

```python
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    import fiftyone as fo
else:
    from .util import try_import_fyftone
    fo = try_import_fiftyone()
```

This system works as intended. If I just use the `from .util import try_import_fyftone
    fo = try_import_fiftyone()` snippet, regardless of `TYPE_CHECKING`, the fiftyone module is not recognized and I cannot type check against it.
    
I was wondering if there's a better way to import a module via a function that is still recognized by pyright ? I am not happy with the `TYPE_CHECKING` thing and alos the fact that the `try_import_fiftyone` function is not checked since it's not called within type checking.

Thanks in advance for the help !

Clément

**Comment by erictraut** [comment, marked as answer]
The return type of `try_import_fiftyone ` is `types.ModuleType`, so you are effectively discarding all type information for this module. If you simply remove the return type annotation for `try_import_fiftyone` (leave it unannotated), pyright will infer the return type based on the return statement.

This trick won't work with mypy because it doesn't infer return types, so if you need this code to work with type checkers other than pyright, then I think your `TYPE_CHECKING` solution is the best approach.

**Reply by ClementPinard** [reply]
thank you for the prompt reply, Indeed it works when removing the return type !

My main concern was autocompletion in vscode, but now it works perfectly, thanks !
